cezo_fl_main.py

```diff
@@ -0,0 +1,122 @@
+from preprocess import preprocess
+
+from cezo_fl.server import CeZO_Server
+from cezo_fl.client import Client
+
+from models.cnn_mnist import CNN_MNIST
+from models.lenet import LeNet
+from models.cnn_fashion import CNN_FMNIST
+from models.lstm import CharLSTM
+from tqdm import tqdm
+from gradient_estimators.random_gradient_estimator import RandomGradientEstimator as RGE
+
+
+def prepare_settings_underseed(args, device):
+    torch.manual_seed(args.seed)
+    if args.dataset == "mnist":
+        model = CNN_MNIST().to(device)
+        criterion = nn.CrossEntropyLoss()
+        optimizer = torch.optim.SGD(
+            model.parameters(), lr=args.lr, weight_decay=1e-5, momentum=args.momentum
+        )
+        # scheduler = torch.optim.lr_scheduler.ExponentialLR(optimizer, gamma=0.8)
+    elif args.dataset == "cifar10":
+        model = LeNet().to(device)
+        criterion = nn.CrossEntropyLoss()
+        optimizer = torch.optim.SGD(
+            model.parameters(), lr=args.lr, weight_decay=5e-4, momentum=args.momentum
+        )
+        # scheduler = torch.optim.lr_scheduler.MultiStepLR(
+        #     optimizer, milestones=[200], gamma=0.1
+        # )
+    elif args.dataset == "fashion":
+        model = CNN_FMNIST().to(device)
+        criterion = nn.CrossEntropyLoss()
+        optimizer = torch.optim.SGD(
+            model.parameters(), lr=args.lr, weight_decay=1e-5, momentum=args.momentum
+        )
+        # scheduler = torch.optim.lr_scheduler.MultiStepLR(
+        #     optimizer, milestones=[200], gamma=0.1
+        # )
+    elif args.dataset == "shakespeare":
+        model = CharLSTM().to(device)
+        criterion = nn.CrossEntropyLoss()
+        optimizer = torch.optim.SGD(
+            model.parameters(), lr=args.lr, momentum=0.9, weight_decay=5e-4
+        )
+        # scheduler = torch.optim.lr_scheduler.MultiStepLR(
+        #     optimizer, milestones=[200], gamma=0.1
+        # )
+
+    if args.grad_estimate_method in ["rge-central", "rge-forward"]:
+        method = args.grad_estimate_method[4:]
+        print(f"Using RGE {method}")
+        grad_estimator = RGE(
+            model,
+            mu=args.mu,
+            num_pert=args.num_pert,
+            grad_estimate_method=method,
+            device=device,
+        )
+    else:
+        raise Exception(
+            f"Grad estimate method {args.grad_estimate_method} not supported"
+        )
+    return model, criterion, optimizer, grad_estimator
+
+
+# def get_warmup_lr(
+#     args, current_epoch: int, current_iter: int, iters_per_epoch: int
+# ) -> float:
+#     overall_iterations = args.warmup_epochs * iters_per_epoch + 1
+#     current_iterations = current_epoch * iters_per_epoch + current_iter + 1
+#     return args.lr * current_iterations / overall_iterations
+
+
+if __name__ == "__main__":
+    args = get_params().parse_args()
+
+    clients = []
+
+    for i in range(args.num_clients):
+        device, train_loader, _ = preprocess(args)
```

Comment:
this is wrong. You need to partition the train loader.

Reply:
Not sure what's the best way to do partition. Remember you and @RogerRogerUSC mentioned something about data distribution, not sure if we need to use that instead

Reply:
There are some related examples in FAST project.

Reply:
Let's leave this for future. 
Checked example from https://github.com/RogerRogerUSC/FAST/blob/main/fashion_main.py, https://github.com/RogerRogerUSC/FAST/blob/main/mnist_main.py, https://github.com/RogerRogerUSC/FAST/blob/main/cifar10_main.py. Find fmnist and mnist are using fedlab while cifar10 implements something on its own.

Reply:
https://pytorch.org/docs/stable/data.html#torch.utils.data.random_split
https://medium.com/codex/a-comprehensive-tutorial-to-pytorch-distributeddataparallel-1f4b42bb1b51
got some concerns using distributedSampler

Reply:
resolved in latest commit